File: src/scalars/forGroupMutationCore.ts
// Task 34's production-unconnected loop mutation primitive. The caller owns
// statement/version/binding identities && the body callback; this module only
// owns iteration frame lifetime && in-place outer-slot carry.

export type LoopMutationSlot<T> = ReadonlyMap<string, T>;

export type ForGroupMutationPlan<Statement> = {
  loopScopeId: string;
  iterationBindingId: string;
  iterationValues: readonly number[];
  generatedStatements: readonly Statement[];
};

export type ForGroupIterationContext<Statement> = {
  loopScopeId: string;
  iterationBindingId: string;
  iterationIndex: number;
  iterationValue: number;
  statement: Statement;
};

export type ForGroupMutationFrame<T> = {
  readonly loopScopeId: string;
  readonly iterationBindingId: string;
  readonly iterationIndex: number;
  readonly iterationValue: number;
  read: (bindingId: string) => T | number | undefined;
  declareLocal: (bindingId: string, value: T) => void;
  set: (bindingId: string, value: T) => void;
};

export type ForGroupMutationEnvironment<T> = {
  run: <Statement>(
    plan: ForGroupMutationPlan<Statement>,
    executeStatement: (
      frame: ForGroupMutationFrame<T>,
      context: ForGroupIterationContext<Statement>
    ) => ForGroupMutationRunOutcome | void
  ) => ForGroupMutationRunOutcome;
  read: (bindingId: string) => T | number | undefined;
  finalSlots: () => ReadonlyMap<string, T>;
};

type ActiveFrame<T> = {
  loopScopeId: string;
  iterationBindingId: string;
  iterationIndex: number;
  iterationValue: number;
  locals: Map<string, T>;
};

export class ForGroupMutationError extends Error {}

/**
 * Existing forGroup expansion supplies this at a generated-statement boundary
 * when its compiled evaluation limit has been reached. The core propagates it
 * through all remaining statements, iterations, && nested runs.
 */
export type ForGroupMutationRunOutcome = "completed" | "stopped";

/**
 * Creates one mutable outer environment. Every iteration gets a fresh local
 * frame, while writes not targeting a local survive into the next iteration.
 * No environment is cloned per iteration.
 */
export const createForGroupMutationEnvironment = <T>(initialSlots: LoopMutationSlot<T>): ForGroupMutationEnvironment<T> => {
  const outerSlots = new Map(initialSlots);
  const frames: ActiveFrame<T>[] = [];

  const frameFor = (): ActiveFrame<T> => {
    const frame = frames.at(-1);
    if (!frame) throw new ForGroupMutationError("forGroup mutation requires an active iteration frame");
    return frame;
  };

  const read = (bindingId: string): T | number | undefined => {
    for (let index = frames.length - 1; index >= 0; index -= 1) {
      const frame = frames[index];
      if (bindingId === frame.iterationBindingId) return frame.iterationValue;
      const local = frame.locals.get(bindingId);
      if (local !== undefined) return local;
    }
    return outerSlots.get(bindingId);
  };

  const set = (bindingId: string, value: T): void => {
    for (let index = frames.length - 1; index >= 0; index -= 1) {
      const frame = frames[index];
      if (bindingId === frame.iterationBindingId) {
        throw new ForGroupMutationError(`forGroup iteration binding ${bindingId} is read-only`);
      }
      if (frame.locals.has(bindingId)) {
        frame.locals.set(bindingId, value);
        return;
      }
    }
    outerSlots.set(bindingId, value);
  };

  const declareLocal = (bindingId: string, value: T): void => {
    const frame = frameFor();
    if (bindingId === frame.iterationBindingId || frame.locals.has(bindingId)) {
      throw new ForGroupMutationError(`forGroup local binding ${bindingId} is already defined`);
    }
    frame.locals.set(bindingId, value);
  };

  const run: ForGroupMutationEnvironment<T>["run"] = (plan, executeStatement) => {
    for (let iterationIndex = 0; iterationIndex < plan.iterationValues.length; iterationIndex += 1) {
      const iterationValue = plan.iterationValues[iterationIndex];
      const active: ActiveFrame<T> = {
        loopScopeId: plan.loopScopeId,
        iterationBindingId: plan.iterationBindingId,
        iterationIndex,
        iterationValue,
        locals: new Map()
      };
      frames.push(active);
      try {
        for (const statement of plan.generatedStatements) {
          const frame: ForGroupMutationFrame<T> = {
            loopScopeId: active.loopScopeId,
            iterationBindingId: active.iterationBindingId,
            iterationIndex: active.iterationIndex,
            iterationValue: active.iterationValue,
            read,
            declareLocal,
            set
          };
          const outcome = executeStatement(frame, {
            loopScopeId: active.loopScopeId,
            iterationBindingId: active.iterationBindingId,
            iterationIndex: active.iterationIndex,
            iterationValue: active.iterationValue,
            statement
          });
          if (outcome === "stopped") return "stopped";
        }
      } finally {
        // Exactly Task 33's frame-lifetime rule: locals disappear at the
        // explicit frame boundary, including when a body callback throws.
        frames.pop();
      }
    }
    return "completed";
  };

  return { run, read, finalSlots: () => new Map(outerSlots) };
};

File: src-tauri/src/evaluation/scalars/for_group_mutation_core.rs
//! Task 34's pure, production-unconnected forGroup mutation runner.
//!
//! The caller supplies opaque binding/version/statement identities and a body
//! callback. This module owns only frame lifetime and in-place outer carry;
//! it deliberately has no IPC, geometry, or evaluator dependency.

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct ForGroupMutationPlan<Statement> {
    pub(crate) loop_scope_id: String,
    pub(crate) iteration_binding_id: String,
    pub(crate) iteration_values: Vec<f64>,
    pub(crate) generated_statements: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum LoopRead<T> {
    Iteration(f64),
    Slot(T),
}

#[derive(Debug, Clone, Copy)]
pub(crate) struct ForGroupIterationContext<'a, Statement> {
    pub(crate) loop_scope_id: &'a str,
    pub(crate) iteration_binding_id: &'a str,
    pub(crate) iteration_index: usize,
    pub(crate) iteration_value: f64,
    pub(crate) statement: &'a Statement,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum ForGroupMutationError {
    NoActiveFrame,
    ReadOnlyIterationBinding(String),
    DuplicateLocalBinding(String),
}

#[derive(Debug)]
struct ActiveFrame<T> {
    loop_scope_id: String,
    iteration_binding_id: String,
    iteration_index: usize,
    iteration_value: f64,
    locals: HashMap<String, T>,
}

/// A stack of Task 33-style lexical frames around one shared outer slot map.
/// Entering/leaving an iteration allocates/drops only the frame's locals.
#[derive(Debug)]
pub(crate) struct ForGroupMutationEnvironment<T> {
    outer_slots: HashMap<String, T>,
    frames: Vec<ActiveFrame<T>>,
}

impl<T: Clone> ForGroupMutationEnvironment<T> {
    pub(crate) fn new(initial_slots: HashMap<String, T>) -> Self {
        Self {
            outer_slots: initial_slots,
            frames: Vec::new(),
        }
    }

    pub(crate) fn read(&self, binding_id: &str) -> Option<LoopRead<T>> {
        for frame in self.frames.iter().rev() {
            if binding_id == frame.iteration_binding_id {
                return Some(LoopRead::Iteration(frame.iteration_value));
            }
            if let Some(value) = frame.locals.get(binding_id) {
                return Some(LoopRead::Slot(value.clone()));
            }
        }
        self.outer_slots
            .get(binding_id)
            .cloned()
            .map(LoopRead::Slot)
    }

    pub(crate) fn declare_local(
        &mut self,
        binding_id: &str,
        value: T,
    ) -> Result<(), ForGroupMutationError> {
        let Some(frame) = self.frames.last_mut() else {
            return Err(ForGroupMutationError::NoActiveFrame);
        };
        if binding_id == frame.iteration_binding_id || frame.locals.contains_key(binding_id) {
            return Err(ForGroupMutationError::DuplicateLocalBinding(
                binding_id.to_owned(),
            ));
        }
        frame.locals.insert(binding_id.to_owned(), value);
        Ok(())
    }

    pub(crate) fn set(&mut self, binding_id: &str, value: T) -> Result<(), ForGroupMutationError> {
        for frame in self.frames.iter_mut().rev() {
            if binding_id == frame.iteration_binding_id {
                return Err(ForGroupMutationError::ReadOnlyIterationBinding(
                    binding_id.to_owned(),
                ));
            }
            if frame.locals.contains_key(binding_id) {
                frame.locals.insert(binding_id.to_owned(), value);
                return Ok(());
            }
        }
        self.outer_slots.insert(binding_id.to_owned(), value);
        Ok(())
    }

    pub(crate) fn final_slots(&self) -> HashMap<String, T> {
        self.outer_slots.clone()
    }

    pub(crate) fn run<Statement, F>(
        &mut self,
        plan: &ForGroupMutationPlan<Statement>,
        mut execute_statement: F,
    ) -> Result<(), ForGroupMutationError>
    where
        F: FnMut(
            &mut Self,
            ForGroupIterationContext<'_, Statement>,
        ) -> Result<(), ForGroupMutationError>,
    {
        for (iteration_index, iteration_value) in plan.iteration_values.iter().copied().enumerate()
        {
            self.frames.push(ActiveFrame {
                loop_scope_id: plan.loop_scope_id.clone(),
                iteration_binding_id: plan.iteration_binding_id.clone(),
                iteration_index,
                iteration_value,
                locals: HashMap::new(),
            });
            let outcome = (|| {
                for statement in &plan.generated_statements {
                    let frame = self
                        .frames
                        .last()
                        .expect("iteration frame must remain active");
                    let loop_scope_id = frame.loop_scope_id.clone();
                    let iteration_binding_id = frame.iteration_binding_id.clone();
                    let iteration_index = frame.iteration_index;
                    let iteration_value = frame.iteration_value;
                    execute_statement(
                        self,
                        ForGroupIterationContext {
                            loop_scope_id: &loop_scope_id,
                            iteration_binding_id: &iteration_binding_id,
                            iteration_index,
                            iteration_value,
                            statement,
                        },
                    )?;
                }
                Ok(())
            })();
            // Match Task 33's frame retirement even on callback failure.
            self.frames.pop();
            outcome?;
        }
        Ok(())
    }
}
